### DISCUSSION TITLE
How to use mbed-ce to run unit tests on the host computer

### DISCUSSION THREAD

**Discussion by ladislas** [mbed-ce/mbed-os#148] [Q&A]
@multiplemonomials in my quest to move from mbed-os to mbed-ce, I'm facing a new issue: running unit tests on the host computer with Google Tests.

With your previous USCRPL/mbed-cmake, I had managed to set everything up and it was working really well.

Now, I can't seem to find a way to do the same.

Using https://github.com/leka/LekaOS you can try it yourself from the branch `ladislas/feature/move-to-mbed-ce`:

```
make config_unit_tests_lite COVERAGE=ON
make ut
```

Relevant CMakeLists.txt is here: https://github.com/leka/LekaOS/blob/ladislas/feature/move-to-mbed-ce/tests/unit/CMakeLists.txt

Is there documentation on how to set that up?

Thanks for your help :)

**Comment by multiplemonomials** [comment]
I will have to be honest, I have not touched the host-side unit test stuff at all since I started working on Mbed CE.  It's exactly as ARM left it.  You should be able to reference the CI workflow that builds the unit tests though, I know that that has whatever's needed to make it work.

**Reply by ladislas** [reply]
> You should be able to reference the CI workflow that builds the unit tests though, I know that that has whatever's needed to make it work.

I was looking for that, but I could not find it 🤔

**Reply by multiplemonomials** [reply]
Ah that's [here](https://github.com/mbed-ce/mbed-os/blob/af6d50aefe923349c32c8a1bd6dd0786ffe66fd7/.github/workflows/basic_checks.yml#L203).  Yeah the naming of that job isn't very clear, huh...

**Comment by ladislas** [comment]
@multiplemonomials I've been making some progress with a lot of trial and errors. I'm also preparing a PR to fix some issues I encountered.

One strange behavior I see is that configuring and compiling for unit tests creates hundreds of `.cmake` and `.a` files related to CTest:

<img width="814" alt="image" src="https://user-images.githubusercontent.com/2206544/223855060-147e7693-297c-49d8-aab3-7ad5d3aa1380.png">

Any ideas how I could avoid that?

**Reply by multiplemonomials** [reply]
I can only assume these are some sort of script used by CTest, though I am not totally sure.

**Comment by ladislas** [comment]
Alright, new issue:

I have a driver target that is linked against `mbed-core-flags` when compiled to be used on the target hardware.

Now, I also have unit test for this drivers. The fact that it is linked against `mbed-core-flags` pulls a lot of dependencies for the unit tests which in turn causes a lot of errors.

In particular, the real `platform/mbed_retarget.h` is pulled instead of the `double` one, which cause a lot of conflicts and errors.

Not sure how to fix this

**Reply by multiplemonomials** [reply]
Hmm, I can see how that would be an issue, yeah.  Apologies, I admit that I have not really tested the UNITTESTS build much other than ensuring I didn't break it...

One workaround would be something like
```cmake
add_library(mbed-headers INTERFACE)
if(UNIT_TEST_MODE)
     target_link_libraries(mbed-headers PUBLIC <various unit test targets>)
else()
     target_link_libraries(mbed-headers PUBLIC mbed-core-flags mbed-rtos-flags)
endif()
```

Then have all your libraries link to `mbed-headers`.  That should work, though I do feel like it's a bit ugly and maybe should be automated away by the build system.

**Reply by ladislas** [reply]
it works! Amazing! :) 

I have two different projets, one for the firmware, the other for the unit tests, so I don't even need to if/else.

do you think this logic should be moved to mbed-ce?

I tried doing the same thing but directly with `mbed-core-flags mbed-rtos-flags` but it did not work and there were too many dependencies to change.

**Reply by multiplemonomials** [reply]
> do you think this logic should be moved to mbed-ce?

Ideally, yeah, but I don't think I have time to do that anytime soon
